### PR TITLE
[Design] Design updates to handler methods and binding

### PR DESCRIPTION
**Pull request by rynowak**
Implements #6091, #6072, #6090, #5927

### CODE REVIEW

**Comment by rynowak** [comment]
Not all tests are passing, and needs a few functional tests for the new naming conventions and attribute.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/CompiledPageActionDescriptor.cs [hunk 1]
```diff
@@ -30,15 +30,25 @@ public CompiledPageActionDescriptor(PageActionDescriptor actionDescriptor)
         }
 
         /// <summary>
-        /// Gets or sets the <see cref="TypeInfo"/> of the page.
+        /// Gets the list of handler methods for the page. 
         /// </summary>
-        public TypeInfo PageTypeInfo { get; set; }
+        public IList<HandlerMethodDescriptor> HandlerMethods { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/CompiledPageActionDescriptor.cs]: I made a variety of updates to these descriptors to make them more consistent with controllers

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/HandlerMethodDescriptor.cs [hunk 2]
```diff
@@ -1,25 +1,19 @@
...
+        public MethodInfo MethodInfo { get; set; }
 
         public string HttpMethod { get; set; }
 
-        public StringSegment FormAction { get; set; }
-
-        public HandlerParameterDescriptor[] Parameters { get; set; }
+        public string FormAction { get; set; }
 
-        public bool OnPage { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/HandlerMethodDescriptor.cs]: LEL

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/HandlerParameterDescriptor.cs [hunk 3]
```diff
@@ -1,16 +1,13 @@
...
 
-using System;
 using System.Reflection;
 using Microsoft.AspNetCore.Mvc.Abstractions;
 
 namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
 {
     public class HandlerParameterDescriptor : ParameterDescriptor
     {
-        public object DefaultValue { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/HandlerParameterDescriptor.cs]: This isn't the right level to handle default values. This should be part of the executor like in MVC.

The general flow is that model binding makes a dictionary, and then we 'splat' the dictionary into an array and the executor will fill in the gaps.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/HandlerMethodDescriptor.cs [hunk 4]
```diff
@@ -1,25 +1,19 @@
...
     public class HandlerMethodDescriptor
     {
-        public MethodInfo Method { get; set; }
-
-        public Func<object, object[], Task<IActionResult>> Executor { get; set; }
+        public MethodInfo MethodInfo { get; set; }
 
         public string HttpMethod { get; set; }
 
-        public StringSegment FormAction { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/HandlerMethodDescriptor.cs]: I don't know what I was smoking when I made this `StringSegment` 🌱

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs [hunk 5]
```diff
@@ -20,21 +23,210 @@ public DefaultPageLoader(RazorCompiler compiler)
 
         public CompiledPageActionDescriptor Load(PageActionDescriptor actionDescriptor)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs]: The design of this is greatly improved. All the reflection is here and data flows like `reflection -> descriptor -> thing doer`. The system isn't so coupled.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs [hunk 6]
```diff
@@ -20,21 +23,210 @@ public DefaultPageLoader(RazorCompiler compiler)
...
+                // Valid names look like:
+                //  - OnGet
+                //  - OnPost
+                //  - OnFooBar
+                //  - OnTraceAsync
+                //  - OnPostEditAsync
+
+                var start = "On".Length;
+                var length = method.Name.Length - start;
+                if (method.Name.EndsWith("Async", StringComparison.Ordinal))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs]: Reminder for me, I need to check the length again after 'trimming' async.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs [hunk 7]
```diff
@@ -20,21 +23,210 @@ public DefaultPageLoader(RazorCompiler compiler)
...
+                }
+
+                var handlerNameStart = start + 1; // Verb must be at least one character.
+                for (; handlerNameStart < length; handlerNameStart++)
+                {
+                    if (char.IsUpper(method.Name[handlerNameStart]))
+                    {
+                        break;
+                    }
+                }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs]: Yeah this is based on casing now. Come at me bro

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs [hunk 8]
```diff
@@ -20,21 +23,210 @@ public DefaultPageLoader(RazorCompiler compiler)
...
+            {
+                var property = properties[i];
+                var bindingInfo = BindingInfo.GetBindingInfo(property.Property.GetCustomAttributes());
+
+                if (bindingInfo == null && bindPropertyOnType == null)
+                {
+                    continue;
+                }
+                
+                var bindPropertyOnProperty = property.Property.GetCustomAttribute<BindPropertyAttribute>();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs]: The `BindProperty` will be looked at by `BindingInfo` so we **will** get here if the property has one of those.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 9]
```diff
@@ -379,8 +379,18 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
...
                 var arguments = await GetArguments(handler);
 
-                var executor = handler.Executor;
-                result = await executor(handler.OnPage ? _page : _model, arguments);
+                Func<object, object[], Task<IActionResult>> executor = null;
+                for (var i = 0; i < actionDescriptor.HandlerMethods.Count; i++)
+                {
+                    if (object.ReferenceEquals(handler, actionDescriptor.HandlerMethods[i]))
+                    {
+                        executor = CacheEntry.Executors[i];
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: The cache now stores an executor for each handler method.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 10]
```diff
@@ -406,7 +416,10 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
                     parameter,
                     value: null);
 
-                arguments[i] = result.IsModelSet ? result.Model : parameter.DefaultValue;
+                if (result.IsModelSet)
+                {
+                    arguments[i] = result.Model;
+                }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: I think this is causing tests to blow up. I'm going to hardcode what we had before here even though it's not 100% right and log a bug for the follow up.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/NonHandlerAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+using System;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages
+{
+    /// <summary>
+    /// Specifies that the targeted method is not a page handler method.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
+    public class NonHandlerAttribute : Attribute
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/NonHandlerAttribute.cs]: I haven't identified a place to put one of these yet, but we always end up needing things like this.

**Comment by ryanbrandenburg** [comment]
Seems cleaner, I'll be interested to hear the full story on Monday.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/CompiledPageActionDescriptor.cs [hunk 12]
```diff
@@ -30,15 +30,25 @@ public CompiledPageActionDescriptor(PageActionDescriptor actionDescriptor)
         }
 
         /// <summary>
-        /// Gets or sets the <see cref="TypeInfo"/> of the page.
+        /// Gets the list of handler methods for the page.
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/CompiledPageActionDescriptor.cs]: Nitpick, space at the end of this.

**Comment by rynowak** [comment]
@pranavkm - updated with unit tests

Working on functional tests now

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs [hunk 13]
```diff
@@ -20,21 +23,210 @@ public DefaultPageLoader(RazorCompiler compiler)
...
+
+            for (var i = 0; i < methods.Length; i++)
+            {
+                var method = methods[i];
+                if (!IsValidHandlerMethod(method))
+                {
+                    continue;
+                }
+
+                if (method.IsDefined(typeof(NonHandlerAttribute)))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs]: yep.

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs]: Move these two checks inside the `IsValidHandlerMethod` (to be consistent with what we do in https://github.com/aspnet/Mvc/blob/0cee00aae1a4064aa2ab2608015dac92d6389c2e/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultApplicationModelProvider.cs#L381

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageBoundPropertyDescriptor.cs [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    public class PageBoundPropertyDescriptor : ParameterDescriptor
+    {
+        public PropertyInfo Property { get; set; }
+
+        public bool SupportsGet { get; set; }
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageBoundPropertyDescriptor.cs]: `BindOnGet`?

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PagePropertyBinderFactoryTest.cs [hunk 15]
```diff
@@ -235,161 +272,173 @@ public void GetModelBinderFactory_ReturnsNullIfPageModelHasNoSettableBoundProper
...
 
             // Assert
-            Assert.Equal(defaultValue, model.PropertyWithDefaultValue);
+            Assert.Equal("value", model.SupportsGet);
+            Assert.Null(model.Default);
         }
 
         [Fact]
-        public async Task ModelBinderFactory_OverwritesDefaultValue()
+        public async Task ModelBinderFactory_BindsPropertyWithoutSupportsGet_WhenRequestIsNotGet()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PagePropertyBinderFactoryTest.cs]: `ModelBinderFactory_BindsPropertiesRegardlessOfSupportsGet_WhenRequestIsNotGet`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageBoundPropertyDescriptor.cs (hunk 14 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageBoundPropertyDescriptor.cs]: Nope. We already agreed on the name in the attribute.

**Pull request closed without merging**